// Compile: protoc --cpp_out=. remote.proto

syntax = "proto3";

package prometheus;

import "types.proto";

// WriteRequest represents a request to write time series data.
message WriteRequest {
  repeated TimeSeries timeseries = 1;
  // Cortex uses this field to determine the source of the write request.
  // We reserve it to avoid any compatibility issues.
  reserved 2;
  repeated MetricMetadata metadata = 3;
}

// ReadRequest represents a remote read request.
message ReadRequest {
  repeated Query queries = 1;

  enum ResponseType {
    // Server will return a single ReadResponse message with matched series that includes a list of raw samples.
    SAMPLES = 0;

    // Server will stream a delimited ChunkedReadResponse message that contains XOR or HISTOGRAM encoded chunks for a single series.
    STREAMED_XOR_CHUNKS = 1;
  }

  // accepted_response_types allows negotiating the content type of the response.
  repeated ResponseType accepted_response_types = 2;
}

// ReadResponse is a response when response_type equals SAMPLES.
message ReadResponse {
  // Results are in the same order as the request's queries.
  repeated QueryResult results = 1;
}

// Query represents a single query in a ReadRequest.
message Query {
  int64 start_timestamp_ms = 1;
  int64 end_timestamp_ms = 2;
  repeated LabelMatcher matchers = 3;
  ReadHints hints = 4;
}

// QueryResult represents the result of a single query.
message QueryResult {
  // Samples within a time series must be ordered by time.
  repeated TimeSeries timeseries = 1;
}

// ChunkedReadResponse is a response when response_type equals STREAMED_XOR_CHUNKS.
message ChunkedReadResponse {
  repeated ChunkedSeries chunked_series = 1;

  // query_index represents the index of the query from ReadRequest.queries these chunks relate to.
  int64 query_index = 2;
}